reconcile/utils/gitlab_api.py

```diff
@@ -398,61 +401,72 @@ def get_merge_request_comments(
             )
         return comments
 
-    def delete_gitlab_comment(self, note: ProjectMergeRequestNote) -> None:
+    @staticmethod
+    def delete_comment(note: ProjectMergeRequestNote) -> None:
         gitlab_request.labels(integration=INTEGRATION_NAME).inc()
         note.delete()
 
-    def delete_merge_request_comments(self, mr_id: int, startswith: str) -> None:
-        comments = self.get_merge_request_comments(mr_id)
+    def delete_merge_request_comments(
+        self,
+        merge_request: ProjectMergeRequest,
+        startswith: str,
+    ) -> None:
+        comments = self.get_merge_request_comments(merge_request)
         for c in comments:
             body = c["body"] or ""
             if c["username"] == self.user.username and body.startswith(startswith):
-                self.delete_gitlab_comment(c["note"])
+                self.delete_comment(c["note"])
 
-    def add_comment_on_merge_request(
-        self, merge_request: ProjectMergeRequest, comment: str
-    ) -> None:
-        gitlab_request.labels(integration=INTEGRATION_NAME).inc()
-        merge_request.notes.create({"body": comment})
+    def get_project_labels(self) -> Set[str]:
+        return {ln.name for ln in self.get_items(self.project.labels.list)}
 
-    def add_merge_request_comment(self, mr_id, comment):
-        gitlab_request.labels(integration=INTEGRATION_NAME).inc()
-        merge_request = self.project.mergerequests.get(mr_id)
-        gitlab_request.labels(integration=INTEGRATION_NAME).inc()
-        merge_request.notes.create({"body": comment})
-
-    def get_project_labels(self):
-        return [ln.name for ln in self.get_items(self.project.labels.list)]
-
-    def get_merge_request_labels(self, mr_id):
-        gitlab_request.labels(integration=INTEGRATION_NAME).inc()
-        merge_request = self.project.mergerequests.get(mr_id)
-        return merge_request.labels
+    @staticmethod
+    def add_label_to_merge_request(
+        merge_request: ProjectMergeRequest,
+        label: str,
+    ) -> None:
+        # merge_request maybe stale, refresh it to reduce the possibility of labels overwriting
```

Comment:
❤️

Reply:
@geoberle this one is actually very tricky, as gitlab api support add/remove labels in seperate parameters to avoid overwriting https://docs.gitlab.com/ee/api/merge_requests.html#update-mr, but `python-gitlab` doesn't use it https://github.com/python-gitlab/python-gitlab/blob/37328416d87f50f64c9bdbdcb49e9b9a96d2d0ef/gitlab/v4/objects/merge_requests.py#L437, so we don't have much choices here, this just reduce the possibility, if it's really unlucky, this will still overwrite.

Reply:
i think this improves the situation a lot already. sometimes when timing is unfortunate, we could see two integrations fighting over labels.